8351354: Enhance java -XshowSettings:security:tls to show enabled TLS groups and signature algorithms

mpdonova opened the pull request:
In this PR I added TLS groups and signature algorithms to the output of the show settings flag. The values are printed in a single column, like the cipher suites. There can be a lot of values so putting on a single line is ugly. I tried putting them in columns, but it is hard to read.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8351354](https://bugs.openjdk.org/browse/JDK-8351354): Enhance java -XshowSettings:security:tls to show enabled TLS groups and signature algorithms (**Enhancement** - P4)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/24424/head:pull/24424` \
`$ git checkout pull/24424`

Update a local copy of the PR: \
`$ git checkout pull/24424` \
`$ git pull https://git.openjdk.org/jdk.git pull/24424/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 24424`

View PR using the GUI difftool: \
`$ git pr show -t 24424`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/24424.diff">https://git.openjdk.org/jdk/pull/24424.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/24424#issuecomment-2776661341)
</details>

mpdonova commented:
/open

mpdonova commented:
/keepalive

jaikiran commented:
Hello Matthew, In recent times we have been adding more of these configuration debugging into the -XshowSettings option. I'm unusure to what extent this option should be advertized or used for gathering such details of the runtime. Would this be more suitable to be done as a `jcmd` command? Something like a `jcmd <pid> VM.security_configuration`? To me, that feels like a more natural place to be adding support for details like these. More so because I think it's very rare for applications to be launching the JVM with the `-XshowSettings` option.

coffeys commented:
> Hello Matthew, In recent times we have been adding more of these configuration debugging into the -XshowSettings option. I'm unusure to what extent this option should be advertized or used for gathering such details of the runtime. Would this be more suitable to be done as a `jcmd` command? Something like a `jcmd <pid> VM.security_configuration`? To me, that feels like a more natural place to be adding support for details like these. More so because I think it's very rare for applications to be launching the JVM with the `-XshowSettings` option.

It's a fair point Jai. Obtaining such data in a live JVM can be useful. IMO, it's also useful to have it recorded in launcher logs so perhaps both approaches are warranted (maybe a follow on JBS enhancement where jcmd can replay the XshowSettings launcher data ?)

@mpdonova  - do you plan to have some test coverage to track this addition ? Perhaps you could edit the original two test files at https://github.com/openjdk/jdk/commit/2e12a123c9955b7360fd752a6a33d59b6271c5af

mpdonova commented:
> Can you attach a snapshot of the output with these new settings?

<img width="1037" height="997" alt="Screenshot 2025-08-01 at 10 07 32 AM" src="https://github.com/user-attachments/assets/0d2efc34-629e-4510-8de9-3acc068a6446" />

coffeys reviewed (approved):
LGTM

seanjmullan commented:
> > Can you attach a snapshot of the output with these new settings?
> 

I can't see all of the groups or the signature algorithms. Is it truncated?

mpdonova commented:
> > > Can you attach a snapshot of the output with these new settings?
> 
> I can't see all of the groups or the signature algorithms. Is it truncated?

It's not truncated, I just needed to resize the terminal window for a screenshot.
<img width="463" height="909" alt="Screenshot 2025-08-01 at 11 11 11 AM" src="https://github.com/user-attachments/assets/a39b6cd2-2e94-4ba4-a872-75e2cf01b0cc" />

seanjmullan commented:
Thanks for the latest screenshot. I don't think the signature algorithms should be "none". If we can't access the provider-specific defaults, then I think we should omit this information for now. @artur-oracle or @haimaychao can you check this out and see if there is a way to get those defaults? Thanks.

haimaychao commented:
> Thanks for the latest screenshot. I don't think the signature algorithms should be "none". If we can't access the provider-specific defaults, then I think we should omit this information for now. @artur-oracle or @haimaychao can you check this out and see if there is a way to get those defaults? Thanks.

https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/javax/net/ssl/SSLParameters.java#L852
https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/javax/net/ssl/SSLParameters.java#L868

So if we never call `setSignatureSchemes()` on the `SSLParameters`, then `getSignatureSchemes()` will return null. But it doesn't mean no signature schemes are used, and it means underlying provider-specific default signature schemes will be used over the TLS connections. We may use `System.getProperty("jdk.tls.client.SignatureSchemes")`, but it works if the property has been set. It seems no public API we can use to query it directly.

haimaychao commented:
Links to API doc should be:
https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/javax/net/ssl/SSLParameters.java#L729
https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/javax/net/ssl/SSLParameters.java#L745

artur-oracle commented:
> > The code currently being reviewed should only display signature schemes set with `jdk.tls.client.SignatureSchemes` system property.
> 
> Should that be displayed with `-XshowSettings:security:tls` or `-XshowSettings:security:properties`? On the surface, the latter makes more sense to me.

I think we should either implement a public API to provide those signature schemes or not display them at all to avoid any confusion. If someone sets `jdk.tls.client.SignatureSchemes` system property they would sure know about it. That property overrides all other signature schemes for both "signature_algorithms" and "signature_algorithms_cert" extensions.

artur-oracle commented:
> > I think we should either implement a public API to provide those signature schemes or not display them at all to avoid any confusion. If someone sets `jdk.tls.client.SignatureSchemes` system property they would sure know about it. That property overrides all other signature schemes for both "signature_algorithms" and "signature_algorithms_cert" extensions.
> 
> I removed the signature algorithms from this output. If accurate lists aren't generated until TLS handshake, then I don't think there's any reason to print a list here.

Sounds good. I've created [JDK-8366364](https://bugs.openjdk.org/browse/JDK-8366364) to address this problem. Once it's done we can include signature algorithms in the output.

mpdonova commented:
> Please add a release note for this change and update the bug title and description to note that you are not including signature algorithms now - thanks.

I updated the bug and added a release note https://bugs.openjdk.org/browse/JDK-8369225.
